From 219b7dfa71ede9d285be013ddce0f313ed0a164c Mon Sep 17 00:00:00 2001
From: Andrew Davis <afd@ti.com>
Date: Wed, 2 Oct 2024 17:26:34 -0500
Subject: [PATCH 02/27] PENDING: remoteproc: core: Make Remoteproc devices DMA
 capable

Use the parent device's properties to set the new Remoteproc device
DMA settings. This allows DMA buffer attachments to Remoteproc
devices.

Signed-off-by: Andrew Davis <afd@ti.com>
---
 drivers/remoteproc/remoteproc_core.c | 13 +++++++++++++
 include/linux/remoteproc.h           |  1 +
 2 files changed, 14 insertions(+)

diff --git a/drivers/remoteproc/remoteproc_core.c b/drivers/remoteproc/remoteproc_core.c
index f276956f2c5c..8003b44ba222 100644
--- a/drivers/remoteproc/remoteproc_core.c
+++ b/drivers/remoteproc/remoteproc_core.c
@@ -20,6 +20,7 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/device.h>
+#include <linux/dma-map-ops.h>
 #include <linux/panic_notifier.h>
 #include <linux/slab.h>
 #include <linux/mutex.h>
@@ -2486,6 +2487,18 @@ struct rproc *rproc_alloc(struct device *dev, const char *name,
 	rproc->dev.driver_data = rproc;
 	idr_init(&rproc->notifyids);
 
+	/* Make device dma capable by inheriting from parent's capabilities */
+	set_dma_ops(&rproc->dev, get_dma_ops(rproc->dev.parent));
+	if (dma_coerce_mask_and_coherent(&rproc->dev, dma_get_mask(rproc->dev.parent)))
+		dev_warn(&rproc->dev, "Failed to set DMA mask. Trying to continue...\n");
+	rproc->dev.dma_parms = &rproc->dma_parms;
+	/*
+	 * We could use dma_get_max_seg_size(rproc->dev.parent) here but the
+	 * parent is not usually setup correctly either, use full 32bit mask
+	 * for now.
+	 */
+	dma_set_max_seg_size(&rproc->dev, DMA_BIT_MASK(32));
+
 	rproc->name = kstrdup_const(name, GFP_KERNEL);
 	if (!rproc->name)
 		goto put_device;
diff --git a/include/linux/remoteproc.h b/include/linux/remoteproc.h
index b4795698d8c2..4cc680f34d41 100644
--- a/include/linux/remoteproc.h
+++ b/include/linux/remoteproc.h
@@ -585,6 +585,7 @@ struct rproc {
 	u16 elf_machine;
 	struct cdev cdev;
 	bool cdev_put_on_release;
+	struct device_dma_parameters dma_parms;
 	DECLARE_BITMAP(features, RPROC_MAX_FEATURES);
 };
 
-- 
2.39.5 (Apple Git-154)

